BXMSDOC-5387: Create process services doc for Kogito. (#36)
sterobin/kie-docs: doc-content/kogito-docs/src/main/asciidoc/bpmn/ref_bpmn-model-example.adoc

[id='ref_bpmn-model-example_{context}']
= BPMN model example

A typical BPMN business process consists of the following basic components:

* Start events to initiate the process
* Tasks or other steps that are completed as part of the process

> * Connectors to link the process nodes and create a flow chart

Reviewer: perhaps, create a/the sequence flow?

Author: Yes, better. Done.

The `Orders.bpmn2` process in the example describes the steps that need to be followed when ordering items. The process includes a script task for writing debug information and a call activity for invoking a sub-process, using a custom `Order` data object.

.Example `Orders.bpmn2` process
image::kogito/bpmn/bpmn-model-example-orders.png[Image of `Orders.bpmn` example process]

The `Add items` subprocess invokes the following `OrderItems.bpmn2` process, which uses a custom Java service `CalculationService.calculateTotal` and a user task to verify the order.

.Example `OrderItems.bpmn2` process invoked as a subprocess
image::kogito/bpmn/bpmn-model-example-order-items.png[Image of `OrderItems.bpmn` example process]

The `Person.bpmn2` process invokes a Decision Model and Notation (DMN) model in a business rule task to determine customer age, followed by a user task for special handling requirements for children, if applicable.

.Example `Person.bpmn2` process invoked as a subprocess

> image::kogito/bpmn/bpmn-model-example-evaluation.png[Image of `OrderItems.bpmn` example process]

Reviewer: @sterobin this image is from another process, please review.

Author: Ha! I noticed it yesterday when reusing the same in another and hoped to correct it before you saw :) Sorry. Fixed.

> Based on these processes and on application configurations, this example service exposes REST operations to create new orders, to list and delete active orders, and to determine the age of any given person.

Reviewer: ```suggestion
Based on these processes and on application configurations, {PRODUCT} generates a set of REST operations to create new orders, to list and delete active orders, and to determine the age of any given person.
```

Author: Done.

For example, the following REST operations use the endpoint `/orders` to interact with customer orders:

.Example POST request body to create a new order with the given data
[source,json]
----
{
  "approver": "john",
  "order": {
    "orderNumber": "12345",
    "shipped": false
  }
}
----

.Example curl command to create a new order with the given data
[source]
----
curl -d '{"approver" : "john", "order" : {"orderNumber" : "12345", "shipped" : false}}' -H "Content-Type: application/json" -X POST http://localhost:8080/orders
----

.Example curl command to list active orders
[source]
----
curl -X GET http://localhost:8080/orders
----

.Example curl command to retrieve an order by ID
[source]
----

> curl -X GET http://localhost:8080/orders/12345

Reviewer: @sterobin should it be more explicit that this is the process id returned by the post or get command?  Just want to make sure people don't get confused with _orderNumber_.

Author: Oops, yes, should. I clarified this with the same example in the other "Creating and running" doc. Have updated and clarified these endpoint examples here. Done.

----

.Example curl command to cancel an order by ID
[source]
----

> curl -X DELETE http://localhost:8080/orders/12345

Reviewer: Same here, regarding id.

Author: Done.

----

> The following is the BPMN source file for `Orders.bpmn2` process model, as an example:

Reviewer: To be honest, I don't see much value in showing the raw xml.

Author: Well, this section serves as a guided tour of a BPMN model example, so just showing what drives it. We do the same with the DMN doc content (show a DMN XML file as an illustration), so am just trying to keep things consistent.

Reviewer: yep. that's fine, thanks.
